Add Exasol connector (#16202)
trinodb/trino: docs/src/main/sphinx/connector/exasol.rst

.. include:: jdbc-case-insensitive-matching.fragment

Querying Exasol
---------------

The Exasol connector provides a schema for every Exasol schema.
You can see the available Exasol schema by running ``SHOW SCHEMAS``::

    SHOW SCHEMAS FROM example;

If you have a Exasol schema named ``web``, you can view the tables
in this schema by running ``SHOW TABLES``::

    SHOW TABLES FROM example.web;

You can see a list of the columns in the ``clicks`` table in the ``web``
schema using either of the following::

    DESCRIBE example.web.clicks;
    SHOW COLUMNS FROM example.web.clicks;

Finally, you can access the ``clicks`` table in the ``web`` schema::

    SELECT * FROM example.web.clicks;

If you used a different name for your catalog properties file, use
that catalog name instead of ``example`` in the above examples.

.. note::
    The Exasol user must have access to the table in order to access it from Trino.
    The user configuration in the connection properties file, determines your
    privileges in these schemas.

.. _exasol-type-mapping:

Type mapping
------------

Because Trino and Exasol each support types that the other does not, this

> connector :ref:`modifies some types <type-mapping-overview>` when reading or

Author: It is defined in file `docs/src/main/sphinx/language/types.rst`

    - ``CHAR(n)``
    -
  * - ``DATE``
    - ``DATE``
    -
  * - ``TIMESTAMP``
    - ``TIMESTAMP``
    -

No other types are supported.

Trino to Exasol type mapping
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Trino supports creating tables with the following types in an Exasol database.
The table shows the mappings from Trino to Exasol data types:

.. list-table:: Trino to Exasol Type Mapping
  :widths: 30, 25, 50
  :header-rows: 1

  * - Trino type
    - Exasol database type
    - Notes
  * - ``TINYINT``
    - ``DECIMAL(3, 0)``
    -
  * - ``SMALLINT``
    - ``DECIMAL(9, 0)``
    -
  * - ``INTEGER``
    - ``DECIMAL(18, 0)``
    -
  * - ``BIGINT``
    - ``DECIMAL(36 ,0)``
    -
  * - ``DECIMAL(p, s)``
    - ``DECIMAL(p, s)``
    -
  * - ``REAL``
    - ``DOUBLE PRECISION``
    -
  * - ``DOUBLE``
    - ``DOUBLE PRECISION``
    -
  * - ``VARCHAR``
    - ``VARCHAR(2000000)``
    -
  * - ``VARCHAR(n)``
    - ``VARCHAR(n)``
    - See :ref:`exasol character mapping`
  * - ``CHAR(n)``
    - ``CHAR(n)``
    - See :ref:`exasol character mapping`
  * - ``DATE``
    - ``DATE``
    -
  * - ``TIMESTAMP``
    - ``TIMESTAMP``
    -
  
No other types are supported.

.. _exasol number mapping:

> Mapping numeric types

Author: Added note "vice versa"

^^^^^^^^^^^^^^^^^^^^^

> An Exasol ``DECIMAL(p, s)`` maps to Trino's ``DECIMAL(p, s)`` except in these

Author: No, it's called `DECIMAL` like in Exasol.
There will be an exception if something goes wrong.

conditions:

- No precision is specified for the column (example: ``DECIMAL`` or
  ``DECIMAL(*)``).
- Scale (``s`` ) is greater than precision.
- Precision (``p`` ) is greater than 36.
- Scale is negative.

.. _exasol character mapping:

> Mapping character types

Author: Added note "vice versa"

> ================

Reviewer: Will do ..

The Exasol connector allows querying and creating tables in an external Exasol
database.

Requirements
------------

To connect to Exasol, you need:

* Exasol database version 7.1 or higher.
* Network access from the Trino coordinator and workers to Exasol.
  Port 8563 is the default port.

Configuration
-------------

To configure the Exasol connector as the ``example`` catalog, create a file
named ``example.properties`` in ``etc/catalog``. Include the following
connection properties in the file:

.. code-block:: text

    connector.name=exasol
    connection-url=jdbc:exa:exasol.example.com:8563
    connection-user=user
    connection-password=secret

The ``connection-url`` defines the connection information and parameters to pass
to the JDBC driver. See the `Exasol JDBC driver documentation
<https://docs.exasol.com/db/latest/connect_exasol/drivers/jdbc.htm#ExasolURL>`_
for more information.

The ``connection-user`` and ``connection-password`` are typically required and
determine the user credentials for the connection, often a service user. You can
use :doc:`secrets </security/secrets>` to avoid using actual values in catalog
properties files.

.. note::

>     If your Exasol database uses a self-signed TLS certificate you will need to

Reviewer: ```suggestion
    If your Exasol database uses a self-signed TLS certificate you must
```

except in these conditions:

- No precision is specified for the column (example: ``DECIMAL`` or
  ``DECIMAL(*)``).
- Scale (``s`` ) is greater than precision.
- Precision (``p`` ) is greater than 36.
- Scale is negative.

.. _exasol character mapping:

Mapping character types
^^^^^^^^^^^^^^^^^^^^^^^

Trino's ``VARCHAR(n)`` maps to ``VARCHAR(n)`` and vice versa if ``n`` is no greater
than 2000000. Exasol does not support longer values.
If no length is specified, the connector uses 2000000.

Trino's ``CHAR(n)`` maps to ``CHAR(n)`` and vice versa if ``n`` is no greater than 2000.
Exasol does not support longer values.

.. include:: jdbc-type-mapping.fragment

.. _exasol-sql-support:

SQL support
-----------

The connector provides read access and write access to data and metadata in
Exasol. In addition to the :ref:`globally available <sql-globally-available>`
and :ref:`read operation <sql-read-operations>` statements, the connector
supports the following statements:

* :doc:`/sql/delete`
* :doc:`/sql/truncate`
* :doc:`/sql/create-table`
* :doc:`/sql/drop-table`
* :doc:`/sql/alter-table`
* :doc:`/sql/comment`

SQL INSERT
^^^^^^^^^^

> Please note that ``INSERT`` and ``CREATE TABLE AS`` is not supported because

Reviewer: ```suggestion
``INSERT`` and ``CREATE TABLE AS`` are not supported because
```

this would require `parallel connections with the Exasol JDBC driver

> <https://exasol.my.site.com/s/article/Parallel-connections-with-JDBC?language=en_US>`_

Reviewer: Actually .. now reading this .. the whole SQL Insert section should be removed .. this is irrelevant to users.

However ultimately we should figure out how to work around this..

which cannot be implemented with the current architecture of Trino. Limiting the
worker count by setting ``task.writer-count`` to 1 in Trino is also not a viable option as discussed in
`issue #16011 <https://github.com/trinodb/trino/issues/16011>`_.

SQL DELETE
^^^^^^^^^^

If a ``WHERE`` clause is specified, the ``DELETE`` operation only works if the
predicate in the condition clause can be fully pushed down to the data source.

> Exasol does not support predicates ``<``, ``>`` and ``!=`` for the condition of ``DELETE`` statements.

Reviewer: Fix wrapping

> The statement will fail with a database error in these cases.

Reviewer: ```suggestion
The statement fails with a database error in these cases.
```

.. include:: alter-table-limitation.fragment

Table functions
---------------

The connector provides specific :doc:`table functions </functions/table>` to
access Exasol.

.. _exasol-query-function:

``query(varchar) -> table``
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The ``query`` function allows you to query the underlying database directly. It

Reviewer: This whole wording and section might have to be updated ... there are some changes around this that recently got merged.

Author: I am not sure how to change this. I used the Oracle documentation as an example and that contains a similar description.

requires syntax native to Exasol, because the full query is pushed down and
processed in Exasol. This can be useful for accessing native features which are
not available in Trino or for improving query performance in situations where
running a query natively may be faster.

.. include:: polymorphic-table-function-ordering.fragment

As a simple example, query the ``example`` catalog and select an entire table::

    SELECT
      *
    FROM
      TABLE(
        example.system.query(

>           query => 'SELECT

Reviewer: Is there some better example that uses a syntax that is specific to Exasol?
